manim/config.py

```diff
@@ -1,283 +1,406 @@
-    dir_config = {}
-    dir_config["media_dir"] = config["media_dir"] or dirs.MEDIA_DIR
-    dir_config["video_dir"] = config["video_dir"] or dirs.VIDEO_DIR
 
-    if not (config["video_dir"] and config["tex_dir"]):
-        if config["media_dir"]:
-            if not os.path.isdir(dir_config["media_dir"]):
-                os.makedirs(dir_config["media_dir"])
-        if not os.path.isdir(dir_config["media_dir"]):
-            dir_config["media_dir"] = "./media"
+def _init_dirs(config):
+    if not (config["VIDEO_DIR"] and config["TEX_DIR"]):
+        if config["MEDIA_DIR"]:
+            if not os.path.isdir(config["MEDIA_DIR"]):
+                os.makedirs(config["MEDIA_DIR"])
+        if not os.path.isdir(config["MEDIA_DIR"]):
+            config["MEDIA_DIR"] = "./media"
         else:
             print(
-                f"Media will be written to {dir_config['media_dir'] + os.sep}. You can change "
-                "this behavior with the --media_dir flag, or by adjusting dirs.py.,"
+                f"Media will be written to {config['media_dir'] + os.sep}. You can change "
+                "this behavior with the --media_dir flag, or by adjusting manim.cfg."
             )
     else:
-        if config["media_dir"]:
+        if config["MEDIA_DIR"]:
             print(
                 "Ignoring --media_dir, since both --tex_dir and --video_dir were passed."
             )
 
-    dir_config["tex_dir"] = (config["tex_dir"]
-                             or dirs.TEX_DIR
-                             or os.path.join(dir_config["media_dir"], "Tex"))
-    dir_config["text_dir"] = (config["text_dir"]
-                              or dirs.TEXT_DIR
-                              or os.path.join(dir_config["media_dir"], "texts"))
+    # Make sure all folders exist
+    for folder in [config["VIDEO_DIR"], config["TEX_DIR"], config["TEXT_DIR"]]:
+        if not os.path.exists(folder):
+            os.makedirs(folder)
 
-    if not config["video_dir"] or dirs.VIDEO_DIR:
-        dir_config["video_dir"] = os.path.join(dir_config["media_dir"], "videos")
 
-    for folder in [dir_config["video_dir"], dir_config["tex_dir"], dir_config["text_dir"]]:
-        if folder != "" and not os.path.exists(folder):
-            os.makedirs(folder)
+config_filename = 'manim.cfg'
+
+foobar =  os.path.join(os.path.dirname(__file__), config_filename)
+print(foobar)
+config_files = [
+    foobar,
+    os.path.expanduser('~/.{}'.format(config_filename)),
+    os.path.join(os.getcwd(), config_filename),
+    ]
+
+prog = os.path.split(sys.argv[0])[-1]
+if prog in ['manim', 'manimcm']:
+    # If called as entrypoint, set default config, and override the
+    # defaults using CLI arguments
+    args = _parse_cli()
+
+    print(args)
+
+    if args.config_file is not None:
+        config_files.append(args.config_file)
+    config, config_parser = _parse_config(args.file, config_files)
+    _update_config_with_args(config, config_parser, args)
```

Comment:
It's pretty confusing to build part of the config in `_parse_config()` and the rest of it in `_update_config_with_args()`. WDYT of first calling `_parse_cli()` and `ConfigParser.read()`, then calling something like `_compile_config(config_parser, args)`?

Reply:
I agree the current setup is confusing. This is one of the main reasons I opened this PR while still WIP: to get y'all's thoughts on this point.

The main problem is that using `ConfigParser` to read the `.cfg` file has an entirely different logic than using `argparse` to read the CLI arguments. There are two main reasons for this:
1. As mentioned before, `ConfigParser` only allows strings, so I tried to put all that type-conversion logic in a single place.
2. Some CLI flags control more than one config variable. For example, the `-t` or `--transparent` flag changes the values of all of `PNG_MODE`, `MOVIE_FILE_EXTENSION` and `BACKGROUND_OPACITY`. Currently, the way this is handled is that the presence of the `-t` flag uses the `[transparent]` section of the `.cfg` file to override the defaults. I think that's clean because in this case the user can change both the actual defaults but also what the `-t` flag does in a single file and without having to write a single line of code.

Reply:
The fact that the two are parsed differently doesn't prevent building the config in one place; it's just a matter of refactoring what's already here, unless I'm missing something?

Reply:
You're right, this is fixed in the latest commit.